### PR TITLE
Enhancement of Shift patient Tab

### LINKED ISSUE #9355: Enhancement of shift patient tab in patient details page
Proposed designs of Shift patient tab

Image
Image
Image

### PR DESCRIPTION
**Pull request by Mahendar0701**
## Proposed Changes

- Fixes #9355 
- Enhancement of Shift patient tab in patient details page
- Enhancement of Shifting tab list view and shift details page

Proposed design : https://www.figma.com/design/Z93EYKSa1MdBmXndsMBJSQ/Care?node-id=3648-1593&p=f&t=74l8V0bymOcWvqdg-0

Shift Patient tab
image

Create Shift form as slideover
image

Update Shift form as slide over
image

List view in shifting
image

Shift Details Page
image

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA

## Summary by CodeRabbit

- **New Features**
	- Enhanced end-to-end tests for patient consultations, including new visibility checks and verification steps.
	- Added localization entries for ambulance details, current shifting status, and comments.
	- Introduced new routing for patient shifting details and referral letters.
	- Implemented a slide-over interface for shift creation and updates.
	- Added a new `ReferralLetter` component for generating patient referral letters.
	- Enhanced the `ShiftingTable` component with a dropdown menu for shift actions and improved status indicators.
	- Introduced new components for form management and validation, including `Form`, `FormField`, and `Textarea`.

- **Bug Fixes**
	- Improved handling of form submissions and navigation paths after consultations.

- **Style**
	- Updated UI components for better layout and visual consistency, including badges and dropdown menus.
	- Enhanced styling for the `SlideOver` component and other UI elements.

- **Chores**
	- Removed unused imports and streamlined existing components for better maintainability.
	- Added new dependencies to enhance form handling and UI capabilities.

### CODE REVIEW

File: src/components/Patient/ShiftCreate.tsx [hunk 1]
```diff
@@ -261,142 +263,138 @@ export const ShiftCreate = (props: patientShiftProps) => {
...
-          required
-          types={["mobile", "landline"]}
-        />
-
-        {careConfig.wartimeShifting && (
-          <div>
-            <FieldLabel required>
-              Name of shifting approving facility
-            </FieldLabel>
+      <Card className="mx-auto mt-4 flex flex-col ">
```
Comment (Jacobjeevan) [inline src/components/Patient/ShiftCreate.tsx]: Using card here doesn't make sense anymore. Adjust the styles to match figma 👍

File: src/components/Patient/ShiftCreate.tsx [hunk 2]
```diff
@@ -261,142 +263,138 @@ export const ShiftCreate = (props: patientShiftProps) => {
...
+        <div className="lg:grid gap-4 lg:grid-cols-2 ">
+          <TextFormField
```
Comment (Jacobjeevan) [inline src/components/Patient/ShiftCreate.tsx:281]: Should switch over to using react hook forms and shadcn; they are being added in `form-field-v1` branch; check it out for examples.

File: src/components/Shifting/ShiftDetails.tsx [hunk 3]
```diff
@@ -40,555 +34,247 @@ export default function ShiftDetails(props: { id: string }) {
...
-          </CardContent>
-        </Card>
-      </PrintPreview>
-    );
-  };
 
-  if (loading) {
-    return <Loading />;
-  }
+            <div className="flex flex-col gap-4 mt-4 lg:mt-0 w-full lg:flex-row lg:w-auto">
```
Comment (Jacobjeevan) [inline src/components/Shifting/ShiftDetails.tsx]: image

- Adjust the font sizes of both update status/details and referral letters.
- Adjust alignment.

File: src/components/Shifting/ShiftingCommentsSection.tsx [hunk 4]
```diff
@@ -46,37 +44,50 @@ const CommentSection = (props: CommentSectionProps) => {
...
+            </PaginatedList.WhenEmpty>
+
             <PaginatedList.Items<CommentModel>>
               {(item) => <Comment {...item} />}
             </PaginatedList.Items>
             <div className="flex w-full items-center justify-center">
               <PaginatedList.Paginator hideIfSinglePage />
             </div>
           </div>
+
```
Comment (Jacobjeevan) [inline src/components/Shifting/ShiftingCommentsSection.tsx:74]: image

Comments background need to be a bit darker.

File: src/components/Patient/PatientDetailsTab/ShiftingHistory.tsx [hunk 5]
```diff
@@ -43,31 +47,73 @@ const ShiftingHistory = (props: PatientProps) => {
...
+    : [];
   return (
     <section className="mt-4">
       <div className="flex justify-between items-center">
-        <h2 className="my-4 ml-0 text-2xl font-semibold leading-tight">
-          {t("shifting_history")}
-        </h2>
-        <ButtonV2
-          className=""
+        <div className="flex mb-4 bg-gray-100 w-fit rounded-lg px-2 py-1">
```
Comment (Jacobjeevan) [inline src/components/Patient/PatientDetailsTab/ShiftingHistory.tsx]: Adjust the font sizes of the tabs; Play around with it - match font size of shift patient button perhaps?

File: src/components/Shifting/ShiftingTable.tsx [hunk 6]
```diff
@@ -64,195 +74,236 @@ export default function ShiftingTable(props: {
...
           </div>
         )}
-        <div className="col-span-1 hidden text-left uppercase sm:hidden md:hidden lg:block">
-          {t("contact_info")}
-        </div>
+
         <div className="col-span-1 hidden text-left uppercase sm:hidden md:hidden lg:block">
           {t("consent__status")}
         </div>
+        <div className="col-span-1 px-2 hidden text-left uppercase sm:hidden md:hidden lg:block">
```
Comment (Jacobjeevan) [inline src/components/Shifting/ShiftingTable.tsx]: Remove uppercase for these.

File: src/components/Shifting/ShiftingTable.tsx [hunk 7]
```diff
@@ -64,195 +74,236 @@ export default function ShiftingTable(props: {
...
+                            e.stopPropagation();
+                            e.preventDefault();
+                          }}
+                        >
+                          . . .
+                        </h1>
+                      </DropdownMenuTrigger>
+                      <DropdownMenuContent
+                        align="end"
+                        className="w-full font-medium"
```
Comment (Jacobjeevan) [inline src/components/Shifting/ShiftingTable.tsx:218]: primary color text; match figma

File: src/components/Shifting/ShiftDetailsUpdate.tsx [hunk 8]
```diff
@@ -331,29 +339,40 @@ export const ShiftDetailsUpdate = (props: patientShiftProps) => {
...
-              home_facility === state.form.assigned_facility_object?.id &&
-              USER_TYPES.findIndex((type) => user_type === type) <
-                USER_TYPES.findIndex((type) => type === "DistrictAdmin") &&
-              !["DESTINATION APPROVED"].includes(option.text)
-            }
-            optionSelectedLabel={(option) => option.text}
-            onChange={handleFormFieldChange}
-            className="w-full bg-white md:col-span-1 md:leading-5"
-          />
+          <div className="col-span-2 grid grid-cols-1 gap-4 md:grid-cols-2">
```
Comment (Jacobjeevan) [inline src/components/Shifting/ShiftDetailsUpdate.tsx:350]: Check the cancel/submit button placement and compare to figma.

Overall, do the same for other elements and try to match.

File: src/components/ui/button.tsx [hunk 9]
```diff
@@ -60,4 +64,19 @@ const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
 );
 Button.displayName = "Button";
 
-export { Button, buttonVariants };
+const AuthorizedButton: React.FC<AuthorizedElementProps & ButtonProps> = ({
```
Comment (Jacobjeevan) [inline src/components/ui/button.tsx]: Separate file in components/common and import button there.

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]

File: src/components/Patient/ShiftCreate.tsx (hunk 2 above)
Comment (Mahendar0701) [reply src/components/Patient/ShiftCreate.tsx:281]: @Jacobjeevan You mean like should i replace the `src/components/Form/FormFields` files in my branch with `form-field-v1` ?

Comment (Mahendar0701) [reply src/components/Patient/ShiftCreate.tsx:281]: I have looked into `src/components/Form/FormFields` almost all used files are same

Comment (Jacobjeevan) [reply src/components/Patient/ShiftCreate.tsx:281]: > @Jacobjeevan You mean like should i replace the `src/components/Form/FormFields` files in my branch with `form-field-v1` ?

No I meant switch to using React Hook Forms and Shadcn's form in ShiftCreate.

File: src/components/Common/AuthorizedButton.tsx [hunk 10]
```diff
@@ -1,9 +1,11 @@
-import { Button, ButtonProps } from "@headlessui/react";
+import * as React from "react";
 
 import AuthorizedChild from "@/CAREUI/misc/AuthorizedChild";
 
 import { AuthorizedElementProps } from "@/Utils/AuthorizeFor";
 
+import { Button, ButtonProps } from "../ui/button";
```
Comment (Jacobjeevan) [inline src/components/Common/AuthorizedButton.tsx]: minor thing, but switch to absolute import. Same thing for button import in other files.

File: src/components/Patient/ShiftCreate.tsx [hunk 11]
```diff
@@ -108,7 +110,7 @@ export const ShiftCreate = (props: patientShiftProps) => {
     errors: { ...initError },
   };
 
-  const { data: patientData } = useTanStackQueryInstead(routes.getPatient, {
+  const { data: _patientData } = useTanStackQueryInstead(routes.getPatient, {
```
Comment (Jacobjeevan) [inline src/components/Patient/ShiftCreate.tsx]: Switch to using `useQuery` instead. useTanStackQueryInstead is for legacy fn support, but we are slowly switching to useQuery (which uses tanstack).

Same applies for any post requests (switch to using `useMutate`), like in the POST req in this file.

File: src/components/Shifting/ReferralLetter.tsx [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+import routes from "@/Utils/request/api";
+import useTanStackQueryInstead from "@/Utils/request/useQuery";
+import { formatDateTime, formatPatientAge } from "@/Utils/utils";
+
+interface ReferralLetterProps {
+  id: string;
+}
+export const ReferralLetter = ({ id }: ReferralLetterProps) => {
+  const { data, loading } = useTanStackQueryInstead(routes.getShiftDetails, {
+    pathParams: { id },
```
Comment (Jacobjeevan) [inline src/components/Shifting/ReferralLetter.tsx]: useQuery btw.

File: src/components/Shifting/ShiftingCommentsSection.tsx
Comment (Jacobjeevan) [inline src/components/Shifting/ShiftingCommentsSection.tsx:1]: Switch to using Button in this file.

File: src/components/Shifting/ShiftDetailsUpdate.tsx [hunk 13]
```diff
@@ -286,6 +287,7 @@ export const ShiftDetailsUpdate = (props: patientShiftProps) => {
 
   useTanStackQueryInstead(routes.getShiftDetails, {
     pathParams: { id: props.id },
+    prefetch: props.open,
```
Comment (Jacobjeevan) [inline src/components/Shifting/ShiftDetailsUpdate.tsx]: useQuery 👍

File: src/components/Shifting/ShiftingCommentsSection.tsx [hunk 14]
```diff
@@ -15,18 +17,15 @@ import { formatDateTime, formatName } from "@/Utils/utils";
...
-    };
+    const payload = { comment: commentBox };
     if (!/\S+/.test(commentBox)) {
-      Notification.Error({
-        msg: t("comment_min_length"),
-      });
+      Notification.Error({ msg: t("comment_min_length") });
       return;
     }
     const { res } = await request(routes.addShiftComments, {
```
Comment (Jacobjeevan) [inline src/components/Shifting/ShiftingCommentsSection.tsx]: This is where we should we using `useMutate`.

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]
New comments as well the update I left a few hours ago - switch to using react hook forms 👍 

Good work so far.

File: src/components/Patient/ShiftCreate.tsx (hunk 2 above)
Comment (Mahendar0701) [reply src/components/Patient/ShiftCreate.tsx:281]: Hey @Jacobjeevan , i have done changes related to `useQuery `and `useMutation `tanstack. 

Regarding shadcn's form, the shadcn form and components are present in `form-field-v1`  branch but not present in `develop` branch. should i create files and place the code in `develop `branch from `form-field-v1` branch ?

Is there any example usage i can refer to?

Comment (Jacobjeevan) [reply src/components/Patient/ShiftCreate.tsx:281]: > Hey @Jacobjeevan , i have done changes related to `useQuery `and `useMutation `tanstack.
> 
> Regarding shadcn's form, the shadcn form and components are present in `form-field-v1` branch but not present in `develop` branch. should i create files and place the code in `develop `branch from `form-field-v1` branch ?
> 
> Is there any example usage i can refer to?

Yep, you can refer to `ScheduleExceptionForm` and `ScheduleTemplateForm` for examples. I'll be pushing my own changes to `PatientRegistration` (src/pages/appointments/PatientRegistration) tonight, so you can check that as well.

You can install Shadcn's components for each (Input, Date Picker etc).

**Comment by Mahendar0701** [comment]
@Jacobjeevan is shifting removed from care 🤔 ? Becoz all files related to shifting are missing in the form-fields branch .

**Comment by Jacobjeevan** [comment]
> @Jacobjeevan is shifting removed from care 🤔 ? Becoz all files related to shifting are missing in the form-fields branch .

It's being removed.

**Comment by rithviknishad** [comment]
> @Jacobjeevan is shifting removed from care 🤔 ? Becoz all files related to shifting are missing in the form-fields branch .

yup

**Pull request closed without merging**